### ISSUE TITLE
pb can't use IAnonymous credentials

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8062]
|<img alt="timothy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| timothy reported|
|-|-|
|Trac ID|trac#439|
|Type|enhancement|
|Created|2003-12-17 23:46:03Z|

Attachments:

* [anonymous.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0fd/0fdf6a63ef21d8a712f68633ace5ce08cd1914ac/983c2873179db4c1b8a9af1250659251f73bb5e0.py) (814 bytes) - added by timothy on 2003-12-17 23:46:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__439 439
type__enhancement enhancement
reporter__timothy timothy
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1071704763000000 1071704763000000
changetime__1469915890973247 1469915890973247
version__ 
owner__ 
cc__itamarst cc__warner cc__timothy cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="timothy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|timothy commented|
|-|-|

```
#!html
<pre>
An error occurs in t.s.pb.PBClientFactory.login() when passing t.c.Anonymous
credentials.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Currently PB only supports UsernamePassword credentials. Ideally we'd like to
support more, and have better way of defining what goes over network, rather
than just type of credential (e.g. some might want to use SRP, or send in clear
text since it's SSL but backend requires plaintext password.)

Meanwhile, you can define your own support for this by subclassing
PBClientFactory and PBServerFactory and extending the "protocol" (remote API).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[#2313](https://github.com/twisted/twisted/issues/2313) is a duplicate of this.  When the patch there is applied, this will be resolved as well.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

Actually [#2312](https://github.com/twisted/twisted/issues/2312).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [21124]) Merge pb-login-2312-2

Author: carlosedp, exarkun
Reviewer: glyph, exarkun, therve
Fixes [#8062](https://github.com/twisted/twisted/issues/8062)
Fixes [#2312](https://github.com/twisted/twisted/issues/2312)

Add support for IAnonymous credentials to PBServerFactory and
PBClientFactory.  It is now possible to attempt a cred login
anonymously using these classes.  The cred portal is ultimately
responsible for determining if the login should succeed or fail,
as is the general case.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|GitHub commented|
|-|-|

In changeset d76a49ad530ca1269d9d62eedcc9d787ad38aede
```
#!CommitTicketReference repository="" revision="d76a49ad530ca1269d9d62eedcc9d787ad38aede"
Merge pull request #439 from twisted/8702-rodrigc-imp

Author: rodrigc
Reviewer: alex
Fixes: #8702

Remove use of deprecated "import imp" in runtime.py
```
